src/ArrayInterface.jl

```diff
@@ -543,6 +543,105 @@ function restructure(x::Array,y)
   reshape(convert(Array,y),size(x)...)
 end
 
+"""
+    insert(collection, index, item)
+
+Return a new instance of `collection` with `item` inserted into at the given `index`.
+"""
+Base.@propagate_inbounds function insert(collection, index, item)
+  @boundscheck checkbounds(collection, index)
+  ret = similar(collection, length(collection) + 1)
+  @inbounds for i in indices(ret)
```

Comment:
I would split this into separate loops. The compiler may do so automatically, but I'd rather not rely on it.
Something like:
```julia
@inbounds for i in firstindex(ret):index-1
    ret[i] = collection[i]
end
ret[index] = item
@inbounds for i in index+1:lastindex(ret)
    ret[i] = collection[i-1]
end
```
to make sure it SIMDs well if possible.